### PR TITLE
Do not tag contao.db.tl_module.<id> in articles

### LINKED ISSUE #2814: Do not tag contao.db.tl_module.<id> for articles
**Affected version(s)**

Contao >=4.9 

**Description**

Every generated `Module` tags the response with `contao.db.tl_module.<id>`, even when rendering articles:

https://github.com/contao/contao/blob/712de2476dc6100cad1876da1f22b809f7558d5c/core-bundle/src/Resources/contao/modules/Module.php#L242

This way the `x-cache-tags` of a page displaying article `ID 5` will look like this even though module `ID 5` is not present:
```
contao.db.tl_article.5,[…],contao.db.tl_module.5
```

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2814 
| Docs PR or issue | -

### CODE REVIEW

**Review by aschempp — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]

**Comment by Toflar** [comment]
This looks incorrect to me. Why would the `Module` know about `ModuleArticle`? And what if there are other (legacy) modules, that have the same issue?
The only way to fix this properly would be to extract the logic to a `protected function getResponseCacheTags()` and override this in `ModuleArticle` to return nothing or so.
I know we're talking legacy modules here but still, it's worth fixing correctly imho.

**Review by Toflar — CHANGES_REQUESTED** [review]
This looks incorrect to me. Why would the `Module` know about `ModuleArticle`? And what if there are other (legacy) modules, that have the same issue?
The only way to fix this properly would be to extract the logic to a `protected function getResponseCacheTags()` and override this in `ModuleArticle` to return nothing or so.
I know we're talking legacy modules here but still, it's worth fixing correctly imho.

**Comment by m-vo** [comment]
> The only way to fix this properly would be to extract the logic to a protected function getResponseCacheTags() and override this in ModuleArticle to return nothing or so.

Thought about this as well and and I agree in general… Although I think articles are the only "virtual" modules that are ever rendered without a `tl_module` context, so this might be overkill. :thinking:

**Comment by fritzmg** [comment]
`Module` is a legacy parent class that already provides functionality for specific classes that are extending it. See for example `Module::renderNavigation` which is not used by `Module` itself, but by `ModuleNavigation` and `ModuleSitemap`. So imho it would not be far fetched to implement special handling for the special `ModuleArticle`.

But otherwise creating a `protected function getResponseCacheTags()` (and then overriding it in `Contao\ModuleArticle`) is easy to do 👍

**Comment by leofeyer** [comment]
I am fine either way.

**Comment by aschempp** [comment]
> But otherwise creating a protected function getResponseCacheTags() (and then overriding it in Contao\ModuleArticle) is easy to do

It is. It just won't work with fragments. I'd go with the "simple" version, even though we know it's not "ideal". `ModuleArticle` isn't really a module anyway …

**Comment by fritzmg** [comment]
> It is. It just won't work with fragments.

Why? This does not affect fragments. Fragments tag their response in `AbstractFrontendModuleController::__invoke`.

File: core-bundle/src/Resources/contao/modules/Module.php [hunk 1]
```diff
@@ -236,10 +236,10 @@ public function generate()
 		}
 
 		// Tag the module (see #2137)
-		if (System::getContainer()->has('fos_http_cache.http.symfony_response_tagger'))
+		if (!empty($tags = $this->getResponseCacheTags()) && System::getContainer()->has('fos_http_cache.http.symfony_response_tagger'))
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/modules/Module.php]: Should be the other way around :) We don't need to collect the tags if there's no response tagger.

**Review by Toflar — COMMENTED** [review]
Just a minor nitpick. Then we can get this merged :)

File: core-bundle/src/Resources/contao/modules/Module.php (hunk 1 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/modules/Module.php]: see 708a32aed64d7eaa2834c037ba88adc9c4579138

**Review by Toflar — APPROVED** [review]
Thank you @m-vo <3

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**